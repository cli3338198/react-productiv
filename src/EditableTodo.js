import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {
  const [isEditing, setIsEditing] = useState(false);

  /** Toggle if this is being edited */
  function toggleEdit() {
    setIsEditing((prevState) => !prevState);
  }

  /** Call remove fn passed to this. */
  function handleDelete() { }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave() { }
  console.log("What is todo?",todo)
  return (
    <div className="EditableTodo">

      {isEditing ? (
        <TodoForm handleSave={update} initialFormData={todo}/>
      ) : (
        <div>

          <Todo
            id={todo.id}
            title={todo.title}
            description={todo.description}
            priority={todo.priority}
          />

          <div className="mb-3">
            
            <div className="float-end text-sm-end">
              <button
                className="EditableTodo-toggle btn-link btn btn-sm"
                onClick={toggleEdit}
              >
                Edit
              </button>
              <button
                className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                onClick={handleDelete}
              >
                Del
              </button>
            </div>
            <Todo />
          </div>
        </div>
      )}
    </div>
  );
}

export default EditableTodo;
