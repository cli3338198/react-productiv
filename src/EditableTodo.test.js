import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoApp from "./TodoApp";
import EditableTodo from "./EditableTodo";
import todos from "./_testData_";

describe("works", function () {
  it("renders without crashing", function () {
    render(<EditableTodo todo={todos[0]} />);
  });

  it("matches snapshot", function () {
    const { container } = render(<EditableTodo todo={todos[0]} />);
    expect(container).toMatchSnapshot();
  });

  it("should remove todo when delete clicked", function () {
    const { container } = render(<TodoApp initialTodos={todos.slice(0, 1)} />);
    const removeBtn = container.querySelector(".EditableTodo-delBtn");
    expect(container.querySelector(".EditableTodo")).toBeInTheDocument();
    fireEvent.click(removeBtn);
    expect(container.querySelector(".EditableTodo")).not.toBeInTheDocument();
  });

  it("should show the form when toggle edit", function () {
    const { container } = render(<TodoApp initialTodos={todos.slice(0, 1)} />);

    const editBtn = container.querySelector(".EditableTodo-toggle");
    expect(
      container.querySelector(".EditableTodo > .NewTodoForm")
    ).not.toBeInTheDocument();
    fireEvent.click(editBtn);
    expect(
      container.querySelector(".EditableTodo > .NewTodoForm")
    ).toBeInTheDocument();

    const goBtn = container.querySelector(".NewTodoForm-addBtn");
    fireEvent.click(goBtn);
    expect(
      container.querySelector(".EditableTodo > .NewTodoForm")
    ).not.toBeInTheDocument();
  });

  it("should correctly update a todo", function () {
    const { container, debug } = render(
      <TodoApp initialTodos={todos.slice(0, 1)} />
    );
    const editBtn = container.querySelector(".EditableTodo-toggle");
    fireEvent.click(editBtn);

    const newForm = container.querySelector(".EditableTodo > .NewTodoForm");
    const descVal = newForm.querySelector("#newTodo-description").value;
    const titleVal = newForm.querySelector("#newTodo-title").value;
    const priorityVal = newForm.querySelector("#newTodo-priority").value;

    expect(descVal).toEqual(todos[0].description);
    expect(titleVal).toEqual(todos[0].title);
    expect(Number(priorityVal)).toEqual(todos[0].priority);

    const desc = newForm.querySelector("#newTodo-description");
    fireEvent.change(desc, { target: { value: "new description" } });

    const goBtn = container.querySelector(".NewTodoForm-addBtn");
    fireEvent.click(goBtn);

    expect(
      container.querySelector(".EditableTodo > .NewTodoForm")
    ).not.toBeInTheDocument();

    // TODO: find good matcher?
    expect(container.querySelector(".Todo")).toContainElement(
      `<div class="Todo" id="${todos[0].id}"><div>
          <b>t1</b> <small>(priority: 2)</small>
        </div>
        <div>
          <small>new description</small>
        </div>
      </div>`
    );
  });
});
