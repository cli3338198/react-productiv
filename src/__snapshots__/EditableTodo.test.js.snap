// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div>
      <div
        class="mb-3"
      >
        <div
          class="float-end text-sm-end"
        >
          <button
            class="EditableTodo-toggle btn-link btn btn-sm"
          >
            Edit
          </button>
          <button
            class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
          >
            Del
          </button>
        </div>
        <div
          class="Todo"
          id="1"
        >
          <div>
            <b>
              t1
            </b>
             
            <small>
              (priority: 
              2
              )
            </small>
          </div>
          <div>
            <small>
              d1
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
